### DISCUSSION TITLE
Registration, Merge and export of 8bit brightfield channel images (.tif)

### DISCUSSION THREAD

**Discussion by andreevg04** [MathOnco/valis#83] [Q&A]
Dear VALIS Community,

I am a programming newbie, but very interested in image analysis. I want to register serial stains on the same piece of tissue in a TMA. At first I segmented the cores and then used color deconvolution to separate the channels and getting a hematoxylin and a DAB channel. I tried using VALIS to register the Hematoxylin and the DAB channels by using this piece of code. 

from valis import registration
slide_src_dir = "/path/to/slides"
results_dst_dir = "./slide_merging_example"  # Registration results saved here
merged_slide_dst_f = "./slide_merging_example/merged_slides.ome.tiff"  # Where to save merged slide

registrar = registration.Valis(slide_src_dir, results_dst_dir)
rigid_registrar, non_rigid_registrar, error_df = registrar.register()

the registration didnt work as it stopped before the Rigid Registration. I also used the simple example for registerring the images but it didnt work.
I have 12 core CPU and 15GB RAM.

**Comment by cdgatenbee** [comment]
Hi @andreevg04,
Sorry you're running into this issue. It sounds like there is some sort of error either with reading the images, or trying to process them. Could you share the error message so I can narrow it down? Also, am I correct in understanding that you'd like to create a multichannel image by merging the deconvolved channels derived by performing stain deconvolution on brightfield IHC images? If so, an alternative approach could be to use valis to align the IHC images, then deconvolve and merge the stain channels using those aligned images (this is the same approach we used to create the 18plex image in Supplemental Figure 4d). If that sounds like it might work, could you test if you are able to align the original TMA cores (i.e. prior to stain deconvolution)? If this isn't quite right, could you maybe describe in a bit more detail what you would like to do, and, if possible, share some images? 

Best,
-Chandler

**Reply by andreevg04** [reply]
Hi Chandler,
you are absolutely correct what the goal is. I tried the alternate approach you suggested by registering the RGB Cores that I extracted. They were then saved as ome.tiffs Unfortunately I wasnt able to fuse them properly in imageJ and somehow it seemed that I always get a bright white image no matter if I invert it or not. I also used grayscale to fuse them as a hyperstack. How did you deconvolved and merged the stain channels afterwards. 
Also is it possible to make the script run through every folder where each of the stainings are saved for every core. 
I am sending you also the log from the errors: the first one is when I stacked the channels together
/home/burgera/.local/lib/python3.10/site-packages/numba/core/decorators.py:262: NumbaDeprecationWarning: numba.generated_jit is deprecated. Please see the documentation at: https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-generated-jit for more information and advice on a suitable replacement.
  warnings.warn(msg, NumbaDeprecationWarning)
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
JVM has been initialized. Be sure to call registration.kill_jvm() or slide_io.kill_jvm() at the end of your script.
==== Converting images

Converting images: 100%|███████████████████████| 3/3 [00:03<00:00,  1.24s/image]

==== Processing images

Processing images : 100%|██████████████████████| 3/3 [00:07<00:00,  2.37s/image]
Normalizing images: 100%|██████████████████████| 3/3 [00:00<00:00,  7.75image/s]
Denoising images  : 100%|██████████████████████| 3/3 [00:00<00:00,  3.99image/s]

==== Rigid registration

Detecting features   : 100%|█████████████████████████████████████████████████████| 3/3 [00:20<00:00,  6.68s/image]
Matching images      : 100%|█████████████████████████████████████████████████████| 3/3 [01:26<00:00, 28.88s/image]
Finding transforms   : 100%|████████████████████████████████████████████████████| 2/2 [00:00<00:00, 520.58image/s]
Finalizing           : 100%|███████████████████████████████████████████████████| 3/3 [00:00<00:00, 6216.85image/s]

======== Rigid registration complete in 2.257 minutes

Killed

The second one is after I ran the channels separately:
/home/burgera/.local/lib/python3.10/site-packages/numba/core/decorators.py:262: NumbaDeprecationWarning: numba.generated_jit is deprecated. Please see the documentation at: https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-generated-jit for more information and advice on a suitable replacement.
  warnings.warn(msg, NumbaDeprecationWarning)
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
JVM has been initialized. Be sure to call registration.kill_jvm() or slide_io.kill_jvm() at the end of your script.

==== Converting images

Converting images: 100%|███████████████████████| 6/6 [00:04<00:00,  1.24image/s]

==== Processing images

Processing images : 100%|██████████████████████| 6/6 [00:12<00:00,  2.15s/image]
Normalizing images: 100%|██████████████████████| 6/6 [00:01<00:00,  4.34image/s]
Denoising images  : 100%|██████████████████████| 6/6 [00:01<00:00,  3.38image/s]

==== Rigid registration

Detecting features   : 100%|███████████████████| 6/6 [00:17<00:00,  2.97s/image]
Matching images      : 100%|█████████████████| 15/15 [00:49<00:00,  3.29s/image]
Finding transforms   : 100%|█████████████████| 5/5 [00:00<00:00, 1533.79image/s]
Finalizing           : 100%|█████████████████| 6/6 [00:00<00:00, 6289.88image/s]
Killed
There are only two folders with mask and processed and no more. 
This is the log (without the whole saving progress) when I ran the basic snipet for the registration from the examples page. 
JVM has been initialized. Be sure to call registration.kill_jvm() or slide_io.kill_jvm() at the end of your script.

==== Converting images


==== Processing images


==== Rigid registration


======== Rigid registration complete in 3.277 minutes


==== Non-rigid registration


======== Non-rigid registration complete in 19.263 seconds


==== Measuring error

saving ./slide_registration_example/registered_slides/1Adeno_8_CK_A-8.ome.tiff (4853 x 4709 and 3 channels)
etc. and it was done. 

About the coding itself we also did an H&E stain, how can this be incorporated in the script that the H&E be registered with the merged/fused pseudo IF Image?

**Comment by cdgatenbee** [comment]
Hi @andreevg04,

Unfortunately, I don't have much experience using ImageJ, so I'm not able to say why the stain deconvolution isn't working as expected. When I needed to do stain deconvolution, I used valis to warp each slide, and then sliced the pyvips.Image into tiles that I could then convert to numpy arrays and perform stain deconvolution on. You can then convert those numpy arrays to pyvips.Images, and stitch them tiles back together to create the full resolution deconvolved image. You would repeat for each of the RGB cores, and merge all of them together to create your pseudo IF image, which you could then save as an ome.tiff. It's a bit of a process to put together, but potentially worth it, especially if you plan run the pipeline on a large number of samples.

It should very much be possible to create a script that you could use to align individual cores. When I have to do something like that, I find it most useful to set up the script to take command line arguments  via `argparse` (things like location of images, where you want to save results, etc...) and then also create a file that will generate a .txt that contains the set of command line arguments for each sample (1 line per sample). 

As long as the H&E image is also a TMA core, it should be also possible to align them along with the rest of other RGB cores. You would just either need to put it in the same folder as the other images, or put all of the image file names you want to align in a list, and then pass that list to the `img_list` argument when you initialize the `Valis` object. 

I hope that answers your questions, but if not, or if you have any others, please don't hesitate to ask.

Best,
-Chandler

**Reply by andreevg04** [reply]
Dear Chandler, thank you for the reply. 
Unfortunately I dont understand much from the arguments etc. since I am super newbie in programming...
Could you assist me with the scripting and a side question to that, aligning the H&E to the same Core how can I use this to train tissue classifiers in QuPath?
I would be very open to talk to you too because I am going to use VALIS a lot but like I said I am struggling to learn how to use it (because like I said programming xp = 0  :( :( )
Best wishes,
Grigor
